# HG002_PAG07162 Long reads
- [HG002_PAG07162 Long reads](#hg002_pag07162-long-reads)
	- [Basecalling](#basecalling)
	- [mapping](#mapping)
		- [minimap2](#minimap2)
		- [QC](#qc)
	- [Small variants call](#small-variants-call)
		- [PMDV](#pmdv)
	- [Small variants call evaluation](#small-variants-call-evaluation)
		- [PMDV](#pmdv-1)
		- [clair3](#clair3)
	- [SV-call](#sv-call)
		- [mmi-Sniffles](#mmi-sniffles)
		- [mmi-cuteSV](#mmi-cutesv)
	- [SV call evaluation](#sv-call-evaluation)
		- [truvari](#truvari)


## Basecalling 
`6.1.7`

```
guppy_basecaller \
-i $FAST5 \
-s $GUPPY \
--records_per_fastq 0 \
-r \
-c dna_r9.4.1_450bps_sup_prom.cfg \
--device 'auto' \
--compress_fastq \
--num_callers 16 \
--chunk_size 1000 \
--gpu_runners_per_device 4 \
--chunks_per_runner 512 \
--disable_pings \
--resume

cat $GUPPY/pass/* > $FASTQ
```

## mapping 
### minimap2 
```
time $minimap2 -t 16 \
-ax map-ont \
$REF \
--MD \
$FASTQ | samtools sort -T '/media/euphrasie/DATA/tmp' -o $BAM 

samtools index $BAM
```

### QC 
see [QC](https://raw.githack.com/ziphra/long_reads/main/HG002/files/HG002_PAG07162_QC.html)

| Alignment summary  | Reads    | Mean coverage | N50   | Median read length | Median identity freq |
|--------------------|----------|---------------|-------|--------------------|----------------------|
| **mmi**            | 9.07E+06 | 63.7          | 19300 | 19300              | 0.970                |


## Small variants call
### PMDV
```
docker run --ipc=host \
	--gpus all \
	-v "${BAMDIR}":"${BAMDIR}" \
	-v "${OUTPUT_DIR}":"${OUTPUT_DIR}" \
	-v "${REF}":"${REF}" \
	kishwars/pepper_deepvariant:r0.8-gpu \
	run_pepper_margin_deepvariant call_variant \
	-o "${OUTPUT_DIR}" \
	-b "${BAM}" \
	-f "${REF}" \
	-p "${OUTPUT_PREFIX}" \
	-t 16 \
	-g \
	--ont_r9_guppy5_sup
  ```

## Small variants call evaluation
### PMDV 
Benchmarked against Hg38 HG002 GiAB small variants truth set.

| **Type**  | **TRUTH.TOTAL** | **TRUTH.TP** | **TRUTH.FN** | **QUERY.TOTAL** | **QUERY.FP** | **QUERY.UNK** | **FP.gt** | **FP.al** | **METRIC.Recall** | **METRIC.Precision** | **METRIC.Frac_NA** | **METRIC.F1_Score** | **TRUTH.TOTAL.TiTv_ratio** | **QUERY.TOTAL.TiTv_ratio** | **TRUTH.TOTAL.het_hom_ratio** | **QUERY.TOTAL.het_hom_ratio ** |
|-----------|-----------------|--------------|--------------|-----------------|--------------|---------------|-----------|-----------|-------------------|----------------------|--------------------|---------------------|----------------------------|----------------------------|-------------------------------|--------------------------------|
| **INDEL** | 525469          | 394601       | 130868       | 863261          | 166108       | 292939        | 24419     | 53062     | 0.75095           | 0.708747             | 0.33934            | 0.729238            |                            |                            | 1.528275734                   | 1.531744641                    |
| **SNP**   | 3365127         | 3357265      | 7862         | 3987375         | 8409         | 621040        | 1024      | 1992      | 0.997664          | 0.997502             | 0.155752           | 0.997583            | 2.100128487                | 1.964344212                | 1.581195853                   | 1.504930637                    |



### clair3

Benchmarked against Hg38 HG002 GiAB small variants truth set.

| Type  | TRUTH.TOTAL | TRUTH.TP | TRUTH.FN | QUERY.TOTAL | QUERY.FP | QUERY.UNK | FP.gt | FP.al | METRIC.Recall | METRIC.Precision | METRIC.Frac_NA | METRIC.F1_Score | TRUTH.TOTAL.TiTv_ratio | QUERY.TOTAL.TiTv_ratio | TRUTH.TOTAL.het_hom_ratio | QUERY.TOTAL.het_hom_ratio |
|-------|-------------|----------|----------|-------------|----------|-----------|-------|-------|---------------|------------------|----------------|-----------------|------------------------|------------------------|---------------------------|---------------------------|
| INDEL | 525469      | 401496   | 123973   | 704652      | 57530    | 235685    | 13663 | 28137 | 0.764072      | 0.877326         | 0.33447        | 0.816792        |                        |                        | 1.52827573414             | 1.6567349806              |
| SNP   | 3365127     | 3359196  | 5931     | 4356352     | 14934    | 981060    | 1071  | 3145  | 0.998238      | 0.995575         | 0.225202       | 0.996905        | 2.10012848676          | 1.82494009926          | 1.58119585325             | 1.62931269243             |

## SV-call
### mmi-Sniffles
```
sniffles -i $BAM \
	--vcf $BASE/vc/sniffles/${SAMPLE}_sniffles.vcf \
	--tandem-repeats $TRF38 \
	--reference $REF \
	-t 16
```
### mmi-cuteSV
```
cuteSV $BAM $REF $CUTESVOUT . \
    --max_cluster_bias_INS 100 \
    --diff_ratio_merging_INS 0.3 \
    --max_cluster_bias_DEL 100 \
    --diff_ratio_merging_DEL 0.3 \
    --threads 16 \
    --genotype
```

## SV call evaluation
### truvari 
|**Ref** |**Aligner**|**Truth set**      |**Caller** |**TP-base**|**TP-call**|**FP** |**FN** |**precision**  |**recall**     |**f1**         |**base cnt**|**call cnt**|**TP-call_TP-gt**|**TP-call_FP-gt**|**TP-base_TP-gt**|**TP-base_FP-gt**|**gt_concordance**|**gt_matrix/(1, 1)/(1, 1)**|**gt_matrix/(1, 1)/(0, 1)**|**gt_matrix/(1, 0)/(0, 1)**|**gt_matrix/(1, 0)/(1, 1)**|**gt_matrix/(0, 1)/(1, 1)**|**gt_matrix/(0, 1)/(0, 1)**|
|----|-------|---------------|-------|-------|-------|---|---|-----------|-----------|-----------|--------|--------|-------------|-------------|-------------|-------------|--------------|-----------------------|-----------------------|-----------------------|-----------------------|-----------------------|-----------------------|
|Hg38|mmi    |CMRG           |cuteSV |194    |194    |16 |22 |0.923809524|0.898148148|0.910798122|216     |210     |137          |57           |137          |57           |0.706185567   |78                     |2                      |37                     |8                      |59                     |10                     |
|Hg38|mmi    |SV tier1 lifted|cuteSV |9180   |9180   |761|390|0.923448345|0.959247649|0.941007637|9570    |9941    |9096         |84           |9096         |84           |0.990849673   |4180                   |47                     |4916                   |37                     |                       |                       |
|Hg38|mmi    |CMRG           |snf    |198    |198    |12 |18 |0.942857143|0.916666667|0.929577465|216     |210     |141          |57           |141          |57           |0.712121212   |81                     |1                      |38                     |9                      |9                      |60                     |
|Hg38|mmi    |SV tier1 lifted|snf    |9290   |9290   |685|280|0.931328321|0.970741902|0.950626759|9570    |9975    |9081         |209          |9081         |209          |0.977502691   |4237                   |48                     |4844                   |159                    |2                      |                       |
