# HG002_PAG07162 Long reads
- [HG002_PAG07162 Long reads](#hg002_pag07162-long-reads)
  - [Basecalling](#basecalling)
  - [mapping](#mapping)
    - [minimap2](#minimap2)
    - [QC](#qc)
  - [SV-call](#sv-call)
    - [mmi-Sniffles](#mmi-sniffles)
  - [SV call evaluation](#sv-call-evaluation)
    - [truvari](#truvari)
  - [Small variants call](#small-variants-call)
    - [PMDV](#pmdv)
  - [Small variants call evaluation](#small-variants-call-evaluation)
  - [SV call evaluation](#sv-call-evaluation-1)
    - [truvari](#truvari-1)


## Basecalling 
`6.1.7`

```
guppy_basecaller \
-i $FAST5 \
-s $GUPPY \
--records_per_fastq 0 \
-r \
-c dna_r9.4.1_450bps_sup_prom.cfg \
--device 'auto' \
--compress_fastq \
--num_callers 16 \
--chunk_size 1000 \
--gpu_runners_per_device 4 \
--chunks_per_runner 512 \
--disable_pings \
--resume

cat $GUPPY/pass/* > $FASTQ
```

## mapping 
### minimap2 
```
time $minimap2 -t 16 \
-ax map-ont \
$REF \
--MD \
$FASTQ | samtools sort -T '/media/euphrasie/DATA/tmp' -o $BAM 

samtools index $BAM
```

### QC 
see [QC](https://raw.githack.com/ziphra/long_reads/main/HG002/files/HG002_PAG07162_QC.html)

| Alignment summary  | Reads    | Mean coverage | N50   | Median read length | Median identity freq |
|--------------------|----------|---------------|-------|--------------------|----------------------|
| **mmi**            | 9.07E+06 | 63.7          | 19300 | 19300              | 0.970                |

## SV-call
### mmi-Sniffles


## SV call evaluation
### truvari 


## Small variants call
### PMDV

## Small variants call evaluation

## SV call evaluation
### truvari 