# HG002 sample ran on PAG07506 flowcell
- [HG002 sample ran on PAG07506 flowcell](#hg002-sample-ran-on-pag07506-flowcell)
  - [Basecalling](#basecalling)
  - [mapping](#mapping)
    - [minimap2](#minimap2)
    - [lra](#lra)
      - [Issue with `lra`](#issue-with-lra)
    - [QC](#qc)
      - [hg38](#hg38)
      - [hg19](#hg19)
      - [T2T](#t2t)
  - [Assembly](#assembly)
    - [Shasta](#shasta)
      - [default parameters](#default-parameters)
      - [low X parameters](#low-x-parameters)
        - [contigs aligned to hg38](#contigs-aligned-to-hg38)
        - [contigs aligned to T2T](#contigs-aligned-to-t2t)
    - [Flye](#flye)
  - [Polishing](#polishing)
    - [medaka](#medaka)
  - [Assembly evaluation](#assembly-evaluation)
    - [hg38](#hg38-1)
    - [T2T](#t2t-1)
  - [SV-call](#sv-call)
    - [lra-cuteSV](#lra-cutesv)
      - [mmi-Sniffles](#mmi-sniffles)


## Basecalling 
`6.1.2`

```
guppy_basecaller \
    -i $FAST5 \
    -s $GUPPY \
    --records_per_fastq 0 \
    -r \
    -c dna_r9.4.1_450bps_hac.cfg \
    --device 'auto' \
    --compress_fastq \
    --num_callers 16 \
    --chunk_size 1000 \
    --gpu_runners_per_device 4 \
    --chunks_per_runner 512 \
    --disable_pings

    cat $GUPPY/pass/* > $FASTQ
```


## mapping 
### minimap2 
```
minimap2 -t 12 \
    -ax map-ont \
    $REF.mmi \
    --MD \
    $FASTQ | samtools sort -o $BAM 
``` 
Memory issues: For some reasons, minimap or samtools right all the tmp files on disk and it will take as much as it is given. 
```
[M::worker_pipeline::40163.976*1.39] mapped 13002 sequences
[E::bgzf_flush] File write failed (wrong size)
[E::bgzf_close] File write failed
samtools sort: failed to create temporary file "/media/euphrasie/DATA/tmp.3308.bam": No space left on device
```

Worked on the 14T Alienware disk.

### lra 
`1.3.2`
```
zcat $FASTQ | lra align -ONT -t 14 $REF - -p s | samtools sort -o $BAM
samtools index $BAM 
```

Concerning memory requirement, `lra` seems more efficient than `minimap2`.

#### Issue with `lra`
Reads with no quality sequence causes PMDV pipeline to crash when DeepVariant runs. See DeepVariant issue [#539](https://github.com/google/deepvariant/issues/539)

Summary of `lra` quality sequence lacking
:
| Sam Flag | Meaning                                         | All       | Q         | No Q    | No Q/Q             |
|----------|-------------------------------------------------|-----------|-----------|---------|--------------------|
| **0**        | mapped to forward strand                        | 1 465 876 | 1 416 309 | 49 567  | 0.0338139105899817 |
| **2048**     | Supplementary alignement                        | 276 887   | 270 331   | 6 556   | 0.0236775291003189 |
| **2064**     | Read reverse strand<br>Supplementary alignment  | 174 337   | 170 082   | 4 255   | 0.0244067524392412 |
| **16**       | Read reverse strand                             | 993 862   | 956 031   | 37 831  | 0.0380646407650157 |
| **4**        | Read unmapped                                   | 604 014   | 440 413   | 163 601 | 0.27085630465519   |

### QC 
#### hg38
see [lra QC](https://rawcdn.githack.com/ziphra/repo/tag/file) and [mmi QC](https://rawcdn.githack.com/ziphra/repo/tag/file).

| Alignment summary  | Reads    | Mean coverage | N50   | Median read length | Median identity freq |
|--------------------|----------|---------------|-------|--------------------|----------------------|
| **mmi**            | 2.14E+06 | 25.3          | 90800 | 15100              | 0.948                |
| **lra**            | 2.01E+06 | 23.6          | 86900 | 16000              | 0.963                |

#### hg19
see [lra QC](https://rawcdn.githack.com/ziphra/repo/tag/file) and [mmi QC](https://rawcdn.githack.com/ziphra/repo/tag/file).

| Alignment summary  | Reads    | Mean coverage | N50   | Median read length | Median identity freq |
|--------------------|----------|---------------|-------|--------------------|----------------------|
| mmi                | 2.14E+06 | 25            | 90200 | 14900              | 0.947                |
| lra                | 2.02E+06 | 23.4          | 87000 | 15300              | 0.962                |

#### T2T
see [lra QC](https://rawcdn.githack.com/ziphra/repo/tag/file) and [mmi QC](https://rawcdn.githack.com/ziphra/repo/tag/file).


| Alignment summary t2t | Reads    | Mean coverage | N50   | Median read length | Median identity freq |
|-----------------------|----------|---------------|-------|--------------------|----------------------|
| mmi                   | 2.14E+06 | 25.5          | 92400 | 15600              | 0.95                 |
| lra                   | 2.01E+06 | 23.9          | 87300 | 17000              | 0.964                |

## Assembly 

### Shasta 
`0.9.0`

#### default parameters
```
shasta \
    --input /home/euphrasie/Documents/HG002_PAG07506/HG002_PAG07506.fastq \
    --config Nanopore-UL-Jan2022 \
    --thread 16 \
    --memoryMode filesystem \
    --memoryBacking disk \
    --Assembly.mode 0 \
    --assemblyDirectory /home/euphrasie/Documents/HG002_PAG07506/ShastaRun
```

The default mode produced excellent assembly results. 2828 Mb of sequence were assembled with an N50 of 38 Mb.The assembly only used 61 Gb of coverage corresponding to only 20 X. 

It is important to remember to run shasta on SSD for performance matters. 


#### low X parameters
```
shasta \
    --input /home/euphrasie/Documents/HG002_PAG07506/HG002_PAG07506.fastq \
    --config Nanopore-UL-Jan2022 \
    --thread 16 \
    --memoryMode filesystem \
    --memoryBacking disk \
    --Assembly.mode 0 \
    --MinHash.minBucketSize 5 \
    --MinHash.maxBucketSize 10 \
    --MarkerGraph.minCoverage 5 \
    --MarkerGraph.maxCoverage 50 \
    --MarkerGraph.edgeMarkerSkipThreshold 50 \
    --assemblyDirectory /home/euphrasie/Documents/HG002_PAG07506/ShastaRun2 
```

##### contigs aligned to hg38 


|        Hg38             | Contigs | N50      | Total length | Read to contig alignment mapping rate | Split read rate | Genome Coverage | Base pair accuracy | QV    | Structural error                                                        |
|---------------------|---------|----------|--------------|---------------------------------------|-----------------|-----------------|--------------------|-------|-------------------------------------------------------------------------|
| Shasta              | 2408    | 38291323 | 2827851270   | 98.92 %                               | 15.28 %         | 90 %            | 0.9974             | 32.60 | 1026:<br>- expansion: 679<br>- collapse: 97<br>- haplotype switch: 250  |
| Shasta x Medaka     | 2408    | 38318697 | 2830396862   | 99.27 %                               | 42.73 %         | 89.84 %         | 0.9981             | 33.74 | 943:<br>- expansion: 188<br>- collapse: 18<br>- haplotype switch: 737   |
| ShastaLowX          | 1728    | 49141234 | 2860162311   | 99.09 %                               | 14.56 %         | 90.41 %         | 0.9974             | 32.60 | 984:<br>- expansion: 643<br>- collapse: 93<br>- haplotype switch: 248   |
| ShastaLowX x Medaka | 1728    | 49179934 | 2862743720   | 99.35 %                               | 42.17 %         | 90.56 %         | 0.9979             | 33.73 | 924:<br>- expansion: 194<br>- collapse: 21<br>- haplotype switch: 709   |


##### contigs aligned to T2T 
|    T2T                 | Contigs | N50      | Total length | Read to contig alignment mapping rate | Split read rate | Genome Coverage | Base pair accuracy | QV    | Structural error                                                        |
|---------------------|---------|----------|--------------|---------------------------------------|-----------------|-----------------|--------------------|-------|-------------------------------------------------------------------------|
| Shasta              | 2408    | 38291323 | 2827851270   | 98.92 %                               | 15.28 %         | 90 %            | 0.9975             | 32.60 | 1026:<br>- expansion: 679<br>- collapse: 97<br>- haplotype switch: 250  |
| Shasta x Medaka     | 2408    | 38318697 | 2830396862   | 99.27 %                               | 42.73 %         | 89.98 %         | 0.9981             | 33.74 | 943:<br>- expansion: 188<br>- collapse: 18<br>- haplotype switch: 737   |
| ShastaLowX          | 1728    | 49141234 | 2860162311   | 99.09 %                               | 14.56 %         | 90.82 %         | 0.9976             | 32.60 | 984:<br>- expansion: 643<br>- collapse: 93<br>- haplotype switch: 248   |
| ShastaLowX x Medaka | 1728    | 49179934 | 2862743720   | 99.35 %                               | 42.17 %         | 91 %            | 0.9981             | 33.73 | 924:<br>- expansion: 194<br>- collapse: 21<br>- haplotype switch: 709   |

### Flye 
`2.9`

```
flye \
    --nano-hq \
    $FASTQ \
    -o $FLYE \
    -g 2.9g \
    --asm-coverage 8 \
    -t 44
```
We do not have sufficient informatics resources to run Flye. 
It would need ~450Gb of RAM and would run for 3 to 4 days on 30 threads.

## Polishing 
### medaka
```
medaka_consensus \
    -i $FASTQ \
    -d $ASSEMBLY/Assembly.fasta \
    -o $MEDAKA \
    -t 16 \
    -m r941_prom_hac_g507
```

## Assembly evaluation
### hg38 

### T2T 

##

## SV-call
### lra-cuteSV
```
cuteSV ${LRA_BAM} $REF ${LRA_CUTESV}.vcf . \
    --max_cluster_bias_INS 100 \
    --diff_ratio_merging_INS 0.3 \
    --max_cluster_bias_DEL 100 \
    --diff_ratio_merging_DEL 0.3 \
    --threads 16
```
#### mmi-Sniffles
Sniffles recquires bam with MD tags. 

```
sniffles -i ${MMI_BAM} \
--vcf ${MMI_SNIFFLES}.vcf \
--tandem-repeats human_GRCh38_no_alt_analysis_set.trf.bed \
--reference $REF \
-t 14 
```
















[clinvar lifted](https://console.cloud.google.com/storage/browser/fc-47de7dae-e8e6-429c-b760-b4ba49136eee/liftover_vcfs/ClinVar?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&authuser=0&prefix=&forceOnObjectsSortingFiltering=false)
